sycl/doc/Assert.md

```diff
@@ -0,0 +1,426 @@
+based approach comes in place. The approach doesn't require any support from
+Device-side Runtime and Native Device Compiler. Neither it does from Low-level
+Runtime.
+
+Within this approach, a dedicated assert buffer is allocated and implicit kernel
+argument is introduced. The argument is an accessor with `discard_read_write`
+or `discard_write` access mode. Accessor metadata is stored to program scope
+variable. This allows to refer to the accessor without modifying each and every
+user's function. Fallback implementation of `__devicelib_assert_fail` restores
+accessor metadata from program scope variable and writes assert information to
+the assert buffer. Atomic operations are used in order to not overwrite existing
+information.
+
+DPCPP Runtime checks contents of the assert buffer for assert failure flag after
+kernel finishes.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+   2. Fallback version of `__devicelib_assert_fail` is called
+   3. Assert information is stored into assert buffer
+   4. Kernel continues running
+ - Host side:
+   1. A distinct thread is launched no later than the point of enqueue of the of
+      kernel with assertions
+   2. This thread polls the enqueued kernels for finish and checks the assert
+      buffer for assert data
+   3. If assert data is present DPCPP Runtime calls `abort()`
+
+
+#### Online-linking fallback `__devicelib_assert_fail`
+
+Online linking against fallback implementation of `__devicelib_assert_fail` is
+performed only when assertion is enabled.
+
+In DPCPP headers one can see if assert is enabled with status of `NDEBUG` macro
+with `#ifdef`'s. This allows to add implicit buffer argument to kernel
+invocation. Here "implicit" means "implicit to the user".
+
+When in DPCPP Runtime Library this knowledge is obtained from device binary
+image descriptor's property sets.
+
+Each device image is supplied with an array of property sets:
+```c++
+struct pi_device_binary_struct {
+  //...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+```
+Each property set is represented by the following struct:
+```c++
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```c++
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+Whenever `isAssertEnabled` property set is present, this specific device image
```

Comment:
Instead of introducing the whole property set with a single property within it, I suggest that we re-use existing "misc" property set and just add an additional property `isAssertEnabled` there. For reference: #3072, #3131

Reply:
A bit below, the text says that no properties are expected in this set. Though, using miscellaneous property set sounds reasonable enough.